File: autocontext/src/autocontext/blobstore/local.py
"""Local filesystem blob store — content-addressed by SHA256 (AC-518)."""

from __future__ import annotations

import hashlib
import shutil
from pathlib import Path
from typing import Any

from autocontext.blobstore.store import BlobStore, prefix_matches, resolve_blob_path


class LocalBlobStore(BlobStore):
    """Content-addressed local filesystem backend.

    Blobs are stored at ``root/<key>`` and their SHA256 digest is
    computed on write. The same content stored under different keys
    will have the same digest but occupy separate files (simplicity
    over dedup for the local backend).
    """

    def __init__(self, root: Path) -> None:
        self.root = root
        self.root.mkdir(parents=True, exist_ok=True)

    def put(self, key: str, data: bytes) -> str:
        digest = _sha256(data)
        path = resolve_blob_path(self.root, key)
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_bytes(data)
        return digest

    def get(self, key: str) -> bytes | None:
        path = resolve_blob_path(self.root, key)
        if not path.is_file():
            return None
        return path.read_bytes()

    def head(self, key: str) -> dict[str, Any] | None:
        path = resolve_blob_path(self.root, key)
        if not path.is_file():
            return None
        data = path.read_bytes()
        return {
            "size_bytes": len(data),
            "digest": _sha256(data),
            "content_type": _guess_content_type(key),
        }

    def list_prefix(self, prefix: str) -> list[str]:
        prefix_path = self.root / prefix.replace("\\", "/")
        base = prefix_path.parent if not prefix_path.is_dir() else prefix_path
        if not base.is_dir():
            return []
        results: list[str] = []
        for path in sorted(base.rglob("*")):
            if path.is_file():
                rel = path.relative_to(self.root).as_posix()
                if prefix_matches(rel, prefix):
                    results.append(rel)
        return results

    def delete(self, key: str) -> bool:
        path = resolve_blob_path(self.root, key)
        if not path.is_file():
            return False
        path.unlink()
        return True

    def put_file(self, key: str, path: Path) -> str:
        dest = resolve_blob_path(self.root, key)
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(str(path), str(dest))
        return _sha256(dest.read_bytes())

    def get_file(self, key: str, dest: Path) -> bool:
        src = resolve_blob_path(self.root, key)
        if not src.is_file():
            return False
        dest.parent.mkdir(parents=True, exist_ok=True)
        shutil.copy2(str(src), str(dest))
        return True


def _sha256(data: bytes) -> str:
    return "sha256:" + hashlib.sha256(data).hexdigest()


def _guess_content_type(key: str) -> str:
    if key.endswith(".json"):
        return "application/json"
    if key.endswith(".ndjson"):
        return "application/x-ndjson"
    if key.endswith(".md"):
        return "text/markdown"
    if key.endswith(".txt"):
        return "text/plain"
    return "application/octet-stream"

File: ts/src/blobstore/local.ts
/** Local filesystem blob store — content-addressed by SHA256 (AC-518). */

import { createHash } from "node:crypto";
import {
  copyFileSync,
  existsSync,
  mkdirSync,
  readFileSync,
  readdirSync,
  statSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { dirname, join, relative } from "node:path";
import {
  type BlobStore,
  type BlobStoreMeta,
  prefixMatches,
  resolveBlobPath,
} from "./store.js";
import { fsError, isNotFoundError } from "./fs-errors.js";

export class LocalBlobStore implements BlobStore {
  #root: string;

  constructor(root: string) {
    this.#root = root;
    try {
      mkdirSync(root, { recursive: true });
    } catch (error) {
      throw fsError("initialize blob store", root, error);
    }
  }

  put(key: string, data: Buffer): string {
    const path = resolveBlobPath(this.#root, key);
    try {
      mkdirSync(dirname(path), { recursive: true });
      writeFileSync(path, data);
      return sha256(data);
    } catch (error) {
      throw fsError("write blob", key, error);
    }
  }

  get(key: string): Buffer | null {
    const path = resolveBlobPath(this.#root, key);
    try {
      return readFileSync(path);
    } catch (error) {
      if (isNotFoundError(error)) return null;
      throw fsError("read blob", key, error);
    }
  }

  head(key: string): BlobStoreMeta | null {
    const path = resolveBlobPath(this.#root, key);
    try {
      const data = readFileSync(path);
      return {
        sizeBytes: data.length,
        digest: sha256(data),
        contentType: guessContentType(key),
      };
    } catch (error) {
      if (isNotFoundError(error)) return null;
      throw fsError("read blob metadata", key, error);
    }
  }

  listPrefix(prefix: string): string[] {
    const results: string[] = [];
    const walk = (dir: string): void => {
      if (!existsSync(dir)) return;
      for (const entry of readdirSync(dir, { withFileTypes: true })) {
        const full = join(dir, entry.name);
        if (entry.isDirectory()) {
          walk(full);
          continue;
        }
        const rel = relative(this.#root, full).replace(/\\/g, "/");
        if (prefixMatches(rel, prefix)) results.push(rel);
      }
    };
    try {
      walk(this.#root);
    } catch (error) {
      if (isNotFoundError(error)) return [];
      throw fsError("list blobs for prefix", prefix, error);
    }
    return results.sort();
  }

  delete(key: string): boolean {
    const path = resolveBlobPath(this.#root, key);
    try {
      unlinkSync(path);
      return true;
    } catch (error) {
      if (isNotFoundError(error)) return false;
      throw fsError("delete blob", key, error);
    }
  }

  putFile(key: string, path: string): string {
    const dest = resolveBlobPath(this.#root, key);
    try {
      mkdirSync(dirname(dest), { recursive: true });
      copyFileSync(path, dest);
      return sha256(readFileSync(dest));
    } catch (error) {
      throw fsError("write blob from file", key, error);
    }
  }

  getFile(key: string, dest: string): boolean {
    const src = resolveBlobPath(this.#root, key);
    try {
      mkdirSync(dirname(dest), { recursive: true });
      copyFileSync(src, dest);
      return true;
    } catch (error) {
      if (isNotFoundError(error)) return false;
      throw fsError("write blob to file", key, error);
    }
  }
}

function sha256(data: Buffer): string {
  return "sha256:" + createHash("sha256").update(data).digest("hex");
}

function guessContentType(key: string): string {
  if (key.endsWith(".json")) return "application/json";
  if (key.endsWith(".ndjson")) return "application/x-ndjson";
  if (key.endsWith(".md")) return "text/markdown";
  if (key.endsWith(".txt")) return "text/plain";
  return "application/octet-stream";
}
